testsuite/ClosedWellTyped.hs

```diff
@@ -0,0 +1,150 @@
+module ClosedWellTyped (
+    ClosedWellTyped(..)
+    ) where
+
+import Control.Applicative as A ((<$>))
+import Control.Monad.State.Lazy
+import Data.Text.Lazy (Text, pack)
+import Test.QuickCheck
+
+import Morte.Core
+
+
+newtype ClosedWellTyped = ClosedWellTyped { unClosedWellTyped :: Expr X }
+    deriving Show
+
+data Env a = Env {
+      bindings :: [(Var, Expr a)]
+    , uniques  :: [Text]
+    }
+
+type GenExpr e a   = StateT (Env e) Gen a
+type GenExprPair e = StateT (Env e) Gen (Expr e,Expr e)
+
+instance Arbitrary ClosedWellTyped where
+    arbitrary = sized rndExpr
+
+
+rndExpr :: Int -> Gen ClosedWellTyped
+rndExpr n = ClosedWellTyped <$> evalStateT genExpr (initEnv n)
+
+initEnv :: Int -> Env a
+initEnv n = Env [] $ map (pack . show) [1..n]
+
+extend :: Var -> Expr e -> GenExpr e ()
+extend x t = modify (\env -> env { bindings = (x,t) : bindings env })
+
+select :: [(Int,GenExprPair e,Env e -> Bool)] -> GenExprPair e
```

Comment:
I would change the type of `select` to `[(Int, Env e -> Maybe (GenExprPair e))] -> GenExprPair e` so that you need less partial functions in the code.